[AmadeusEka] iP

AmadeusEka opened the pull request:
### Sage

> "Do not keep all your knowledge in your mind; share it, write it down, and let your mind be free." - Sage

Sage is here to help you focus on what matters by taking care of all the things you need to remember. It's:

- Text-based
- Easy to learn
- ~Super fast~ EXTREMEMLY FAST to use :fire:

Getting started is simple:
1. Download Sage from [here](https://github.com/AmadeusEka/ip/releases/tag/A-Jar)
2. Run ` java -jar Sage.jar`
3. Add your tasks and let Sage do the rest! 😉

And the best part? It's absolutely **FREE!**

Features: 
- [x]  Managing tasks
- [x] Deadlines 
- [ ] Reminders 

For Java Enthusiasts:
If you're a Java programmer, you can use Sage to practice your Java skills as well. Dive into the source code, explore how it works, or even contribute by building new features! Here's a sneak peek at the main method:

```
public class Sage {
    // Other helper methods

    public static void main(String[] args) {
        new Sage("data/sage.txt").run();
    } 
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

AmadeusEka committed:
feat: Add greeting

AmadeusEka committed:
Add support for echo

AmadeusEka committed:
Add support for add list

AmadeusEka committed:
Add support for marking tasks

AmadeusEka committed:
Add support for tracking three types of tasks: ToDos, Deadlines and Events

AmadeusEka committed:
Add testing via input/output redirection technique

AmadeusEka committed:
Add error handling for mark/todo/deadline/event

AmadeusEka committed:
Add support for deleting tasks

AmadeusEka committed:
improve error message display

AmadeusEka committed:
Add support for loading and saving tasks from/into the hard disk

AmadeusEka committed:
Merge branch 'branch-Level-7'

AmadeusEka committed:
Update deadline task to record and display date and time

AmadeusEka committed:
Rename Memory class to Storage class

AmadeusEka committed:
Fix deadline task date saving format

AmadeusEka committed:
Refactor and abstract user interactions with Ui class

AmadeusEka committed:
Refactor and abstract task list operations into TaskList class

AmadeusEka committed:
Rename Task subclasses name

AmadeusEka committed:
Improve deadline task input date format error handling

AmadeusEka committed:
Refactor and abstract Parser and Command Class

AmadeusEka committed:
Merge branch 'A-MoreOOP'

AmadeusEka committed:
Refactor the classes into suitable java packages

AmadeusEka committed:
Merge branch 'A-Packages'

AmadeusEka committed:
Merge remote-tracking branch 'origin/add-gradle-support'

AmadeusEka committed:
Add dependency for gradle

AmadeusEka committed:
Refactor package file structure

AmadeusEka committed:
Add JUnit tests to test the behavior of delete command and todo task

AmadeusEka committed:
Merge branch 'A-JUnit'

AmadeusEka committed:
Update storage class date file check

If the data file does not exist, storage class will create a new file

AmadeusEka committed:
Add Javadocs

AmadeusEka committed:
Add support for finding keywords in task list

AmadeusEka committed:
Update find command to search for phrases in task list

AmadeusEka committed:
Refactor all classes based on Coding Standards

AmadeusEka committed:
Merge branch 'branch-A-JavaDoc'

AmadeusEka committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/sage/List/TaskList.java
#	src/main/java/sage/Storage.java
#	src/main/java/sage/Task/Task.java

AmadeusEka committed:
Merge branch 'branch-Level-9'

# Conflicts:
#	src/main/java/sage/List/TaskList.java

xinweichong commented on src/main/java/sage/Command/AddCommand.java:
```diff
@@ -0,0 +1,97 @@
+import sage.Task.ToDoTask;
+import sage.Ui;
+
+public class AddCommand extends Command {
```
Perhaps you can add Javadoc for all classes to give a description of what the class is for!

althea28 replied:
I like how clean and organised your code is! The different classes for each type of command makes it very clear 🔥

xinweichong commented on src/main/java/sage/Command/Command.java:
```diff
@@ -0,0 +1,28 @@
+     */
+    public abstract void execute(TaskList tasks, Ui ui, Storage storage) throws SageException;
+
+    public boolean isExit() {
+        return isExit;
+    }
```
Perhaps can differentiate the variable name from its getter to avoid confusion in code!

xinweichong commented on src/main/java/sage/Task/DeadlineTask.java:
```diff
@@ -0,0 +1,40 @@
+public class DeadlineTask extends Task {
+
+    protected LocalDateTime by;
+    private static final DateTimeFormatter INPUT_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd HHmm");
+    private static final DateTimeFormatter OUTPUT_FORMATTER = DateTimeFormatter.ofPattern("MMM d yyyy hh:mm a");
```
Great use of constants to increase readibility of code!

xinweichong commented on src/main/java/sage/Command/DeleteCommand.java:
```diff
@@ -0,0 +1,41 @@
+        this.indexString = indexString;
+    }
+
+    /**
+     * Executes the delete command, removing the task from the task list.
+     * If the index is not a number or is invalid, a SageException is thrown.
+     *
+     * @param tasks   The TaskList from which the task will be deleted.
+     * @param ui      The Ui object to handle user interaction.
+     * @param storage The Storage object for saving changes to the file.
+     * @throws SageException If the index is not a valid number or is out of bounds.
```
Great descriptions in Javadoc that provide explanations and increases readibility and understanding of code throughout!

AmadeusEka committed:
Add Checkstyle to detect coding style violations

AmadeusEka committed:
Merge branch 'A-CheckStyle'

althea28 commented on src/main/java/sage/Ui.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Scanner;
+
+public class Ui {
+    public static final String NAME = "sage";
```
Perhaps can declare this class variable as private and use access functions instead?

althea28 commented on src/main/java/sage/Storage.java:
```diff
@@ -0,0 +1,88 @@
+     * @throws SageException If the file is corrupted or contains invalid data.
+     */
+    public List<Task> loadTasks() throws IOException, SageException {
+        List<Task> taskList = new ArrayList<>();
```
Perhaps can use plural form 'tasks' to represent collection of tasks?

althea28 commented on src/test/java/sage/Command/DeleteCommandTest.java:
```diff
@@ -0,0 +1,59 @@
+package sage.Command;
+
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import sage.List.TaskList;
+import sage.SageException;
+import sage.Storage;
+import sage.Task.ToDoTask;
+import sage.Ui;
```
Perhaps can consider separating the imports from org and sage for clarity!

althea28 reviewed (commented):
I really like how clean and organised your code is! Struggled quite a bit to point out areas for improvement 🔥

AmadeusEka committed:
Add a GUI to the chatbot

Use the JavaFX technology to implement the GUI

AmadeusEka committed:
Merge branch 'branch-Level-10'

AmadeusEka committed:
Add javadocs

Javadocs added to command and sage methods.
Refactoring was also done to adhere to coding standards.

AmadeusEka committed:
Add assertions for task creation/deletion and validation

Incorporated assertions in the `execute` method of `AddCommand` and
`DeleteCommand` to validate task creation and task list modifications.
Added assertions to ensure that tasks are not null for creation.
Added an assertion to verify that the task is added to the task list.
Added an assertion to verify that the task is deleted.

Incorporated assertions in `Storage` to validate file creation if it
does not exists.

AmadeusEka committed:
Package GUI related classes together

GUI classes includes the main, mainwindow, launcher and dialogbox class.

AmadeusEka committed:
Refactor the import statements

AmadeusEka committed:
Add tasktype class

Used enums for task types to improve readability for add command

AmadeusEka committed:
Abstract storage class logic

In accordance to avoiding having multiple levels of abstractions in
one code block and reduce repeated code.

AmadeusEka committed:
Merge branch 'branch-A-Assertions'

AmadeusEka committed:
Merge branch 'master' into branch-A-CodeQuality

AmadeusEka committed:
Merge pull request #1 from AmadeusEka/branch-A-CodeQuality

Branch a code quality

AmadeusEka committed:
Merge branch 'master' of https://github.com/AmadeusEka/ip

AmadeusEka committed:
Add tagging to tasks

AmadeusEka committed:
Update README file

AmadeusEka committed:
Fix README

Removed unnecessary line

AmadeusEka committed:
Merge branch 'branch-A-UserGuide'

AmadeusEka committed:
Rename Ui.png file
